tools/ci_build/rewriter.py

```diff
@@ -0,0 +1,226 @@
+#!/usr/bin/env python3
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import os
+import shutil
+import onnx
+from onnx import AttributeProto as AP
+
+#pylint: disable=no-member,too-many-locals,too-many-statements
+
+def extract_ops_from_file(file_path, referred_ops):
+    '''extract ops from csv file with format:
+        op_type,domain,opset'''
+
+    if not os.path.isfile(file_path):
+        return referred_ops
+
+    with open(file_path, 'r') as csv_to_read:
+
+        for line in csv_to_read.readlines():
+            op_type, domain, raw_opset = line.strip().split(',')
+            opset = int(raw_opset)
+
+            if op_type not in referred_ops:
+                referred_ops[op_type] = {domain:[opset]}
+
+            elif domain not in referred_ops[op_type]:
+                referred_ops[op_type][domain] = [opset]
+
+            elif opset not in referred_ops[op_type][domain]:
+                referred_ops[op_type][domain].append(opset)
+
+    return referred_ops #end of extract_ops_from_file(...)
+
+
+def extract_ops_from_model(model_path, referred_ops):
+    '''extract ops from models under model_path and return a diction'''
+
+    if not os.path.isdir(model_path):
+        return referred_ops
+
+    def map_domain(domain):
+        if domain == 'ai.onnx.ml':
+            return 'kMLDomain'
+        return 'kOnnxDomain'
+
```

Comment:
There are a number of domains to handle.  Ideally those are read from core/graph/constants.h

e.g.
domain_map = {'kOnnxDomain': 'ai.onnx',
              'kMLDomain': 'ai.onnx.ml',
              'kMSDomain': 'com.microsoft',
              'kMSFeaturizersDomain': 'com.microsoft.mlfeaturizers',
              'kMSNchwcDomain': 'com.microsoft.nchwc'}

Reply:
Since we are only dealing with cpu and cuda provider cc files, do have domains other than "kOnnxDomain" and "kMLDomain" there?

Reply:
Need to handle the contrib ops as well. 

---
In reply to: [468919925](https://github.com/microsoft/onnxruntime/pull/4747#discussion_r468919925) [](ancestors = 468919925)

Reply:
You mean at least adding 'kMSDomain': 'com.microsoft' there right?